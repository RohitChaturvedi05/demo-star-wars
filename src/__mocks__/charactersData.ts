import {
    CharacterDetails,
    Characters,
    Film,
    Starship,
} from '../models/Character';

export const mockCharacters: Characters = [
    {
        uid: '1',
        id: '1',
        name: 'Luke Skywalker',
        gender: 'male',
        skin_color: 'fair',
        hair_color: 'blond',
        height: '172',
        eye_color: 'blue',
        mass: '77',
        birth_year: '19BBY',
        homeworldUrl: 'http://swapi.dev/api/planets/1/',
        homeworld: '1',
    },
    {
        uid: '2',
        id: '2',
        name: 'Leia Organa',
        gender: 'female',
        skin_color: 'light',
        hair_color: 'brown',
        height: '150',
        eye_color: 'brown',
        mass: '49',
        birth_year: '19BBY',
        homeworldUrl: 'http://swapi.dev/api/planets/2/',
        homeworld: '2',
    },
];

export const mockFavoriteCharacters: Characters = [
    {
        uid: '1',
        id: '1',
        name: 'Luke Skywalker',
        gender: 'male',
        skin_color: 'fair',
        hair_color: 'blond',
        height: '172',
        eye_color: 'blue',
        mass: '77',
        birth_year: '19BBY',
        homeworldUrl: 'http://swapi.dev/api/planets/1/',
        homeworld: '1',
    },
];

export const mockPlanets = {
    1: {
        id: '1',
        name: 'Tatooine',
        diameter: '10465',
        gravity: '1',
        population: '200000',
    },
    2: {
        id: '2',
        name: 'Alderaan',
        diameter: '12500',
        gravity: '1',
        population: '2000000000',
    },
};

export const mockCharacterDetails: CharacterDetails = {
    uid: '1',
    id: '1',
    name: 'Luke Skywalker',
    gender: 'male',
    skin_color: 'fair',
    hair_color: 'blond',
    height: '172',
    eye_color: 'blue',
    mass: '77',
    birth_year: '19BBY',
    homeworldUrl: 'http://swapi.dev/api/planets/1/',
    homeworld: '1',
    films: ['http://swapi.dev/api/film/1', 'http://swapi.dev/api/film/2'],
    starships: [
        'http://swapi.dev/api/starship/1',
        'http://swapi.dev/api/starship/2',
    ],
};

export const mockFilms: Film[] = [
    {
        id: '1',
        title: 'A New Hope',
        director: 'George Lucas',
        producer: 'Gary Kurtz, Rick McCallum',
        release_date: '1977-05-25',
    },
    {
        id: '2',
        title: 'The Empire Strikes Back',
        director: 'Irvin Kershner',
        producer: 'Gary Kurtz, Rick McCallum',
        release_date: '1980-05-17',
    },
];

export const mockStarship: Starship[] = [
    {
        name: 'Millennium Falcon',
        cargo_capacity: '100000',
        passengers: '6',
        max_atmosphering_speed: '1050',
        manufacturer: 'Corellian Engineering Corporation',
    },
    {
        name: 'X-wing',
        cargo_capacity: '100000',
        passengers: '6',
        max_atmosphering_speed: '1050',
        manufacturer: 'Incom Corporation',
    },
];
