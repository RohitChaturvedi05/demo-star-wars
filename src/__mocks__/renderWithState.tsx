import { render } from '@testing-library/react';
import React from 'react';
import { StateProvider } from '../state';

export const renderWithState = (Component: React.ReactNode) => {
    return render(<StateProvider>{Component}</StateProvider>);
};
