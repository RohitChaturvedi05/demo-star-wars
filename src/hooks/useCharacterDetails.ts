import { useCallback, useMemo, useState } from 'react';
import { Gender } from '../components/Gender';
import { Character, Film, Starship } from '../models/Character';
import { getCharacterById } from '../services/getCharacterById';
import { getFilm } from '../services/getFilms';
import { getPlanetsByIds } from '../services/getPlanets';
import { getStarShip } from '../services/getStarship';
import { useStateContext } from '../state';
import { ActionType } from '../state/actions';

export const useCharacterDetails = () => {
    const [message, setMessage] = useState('');
    const { state, dispatch } = useStateContext();
    const { planets } = state;
    const [character, setCharacter] = useState<Character>();
    const [films, setFilms] = useState<Film[]>([]);
    const [starShips, setStarShips] = useState<Starship[]>([]);

    const getCharacterDetails = useCallback(
        async (id: string) => {
            const characterData = await getCharacterById(id);
            if (characterData) {
                const planetId = parseInt(characterData?.homeworld ?? '');
                if (planets[planetId] === undefined) {
                    const data = await getPlanetsByIds([planetId]);
                    dispatch({ type: ActionType.SET_PLANETS, payload: data });
                }
                const filePromise = characterData?.films.map(getFilm);
                const starShipPromise =
                    characterData?.starships.map(getStarShip);

                const allFilms = await Promise.all(filePromise);
                const allStarShips = await Promise.all(starShipPromise);

                setFilms(allFilms.filter((film) => !!film) as Film[]);
                setStarShips(
                    allStarShips.filter((starShip) => !!starShip) as Starship[]
                );
                setCharacter(characterData);
            }
        },
        [dispatch, planets]
    );

    const onAddToFavorites = useCallback(
        (id: string) => {
            dispatch({
                type: ActionType.SET_FAV_CHARACTERS,
                payload: id,
            });
        },
        [dispatch]
    );

    const filmsNames = useMemo(
        () => films.map((film) => ({ name: film.title })),
        [films]
    );
    const starShipsNames = useMemo(
        () => starShips.map((starShip) => ({ name: starShip.name })),
        [starShips]
    );

    const isFavorite = useMemo(
        () => character && state.favCharacters.includes(character.id),
        [character, state.favCharacters]
    );

    const onGenderChange = (value: Gender, id: string) => {
        dispatch({ type: ActionType.SET_GENDER, payload: { value, id } });
        setTimeout(() => {
            setMessage('Gender Saved !!');
        }, 2000);
    };

    return {
        message,
        films,
        starShips,
        planets,
        character,
        filmsNames,
        isFavorite,
        starShipsNames,
        onGenderChange,
        onAddToFavorites,
        getCharacterDetails,
    };
};
