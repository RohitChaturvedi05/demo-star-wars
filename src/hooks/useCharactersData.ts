import { useCallback, useMemo, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { APP_PATH } from '../routes/constants';
import { getCharacters } from '../services/getCharacters';
import { getPlanetsByIds } from '../services/getPlanets';
import { getCharactersSearch } from '../services/searchByCharacterName';
import { extractPlanetId } from '../services/utils/extractPlanetId';
import { useStateContext } from '../state';
import { ActionType } from '../state/actions';
import { useDebounce } from './useDebounce';

export const useCharactersData = () => {
    const [queryTerm, setQueryTerm] = useState('');
    const debounceQuery = useDebounce(queryTerm, 300);

    const [isLoading, setLoading] = useState(false);
    const { state, dispatch } = useStateContext();
    const { planets } = state;

    const navigate = useNavigate();

    const getCharactersData = useCallback(
        async (page: number) => {
            setLoading(true);
            const { characters, total_pages } = await getCharacters({
                page,
            });

            // get planets ids from characters
            const planetsId = characters
                .map((item) => extractPlanetId(item.homeworldUrl))
                .filter((id): id is number => id !== null && id !== undefined)
                .filter((id) => planets[id] === undefined);

            // Update planets state if needed
            if (planetsId.length > 0) {
                const planetsData = await getPlanetsByIds(
                    Array.from(new Set(planetsId))
                );
                dispatch({
                    type: ActionType.SET_PLANETS,
                    payload: planetsData,
                });
            }
            dispatch({ type: ActionType.SET_CHARACTERS, payload: characters });
            dispatch({ type: ActionType.SET_TOTAL_PAGE, payload: total_pages });
            dispatch({
                type: ActionType.SET_CURRENT_PAGE,
                payload: page,
            });

            setLoading(false);
        },
        [dispatch, planets]
    );

    const searchCharactersByName = useCallback(
        async (name: string) => {
            setLoading(true);
            const { characters } = await getCharactersSearch({
                page: 1,
                name,
            });
            dispatch({ type: ActionType.SET_CHARACTERS, payload: characters });
            dispatch({ type: ActionType.SET_TOTAL_PAGE, payload: 0 });
            setLoading(false);
        },
        [dispatch]
    );

    const onFavoriteToggle = useCallback(
        (id: string) => {
            dispatch({
                type: ActionType.SET_FAV_CHARACTERS,
                payload: id,
            });
        },
        [dispatch]
    );

    const onDetailClick = useCallback(
        (id: string) => {
            navigate(`${APP_PATH.DETAILS}/${id}`);
        },
        [navigate]
    );

    const onPageChange = async (page: number) => {
        dispatch({
            type: ActionType.SET_CURRENT_PAGE,
            payload: page,
        });
        await getCharactersData(page);
    };

    const onSearch = async (searchTerm: string) => {
        setQueryTerm(searchTerm);
        setLoading(true);
        if (searchTerm.length === 0) {
            await onPageChange(1);
        }
    };

    const initializeState = useCallback(
        () => getCharactersData(1),
        [getCharactersData]
    );

    const favorites = useMemo(() => {
        return state.characters.filter(({ uid }) =>
            state.favCharacters.includes(uid)
        );
    }, [state.characters, state.favCharacters]);

    return {
        isLoading,
        favorites,
        queryTerm,
        debounceQuery,
        characters: state.characters,
        planets: state.planets,
        favCharacters: state.favCharacters,
        currentPage: state.currentPage,
        totalPage: state.totalPage,
        initializeState,
        onSearch,
        searchCharactersByName,
        onFavoriteToggle,
        onDetailClick,
        getCharactersData,
        onPageChange,
    };
};
