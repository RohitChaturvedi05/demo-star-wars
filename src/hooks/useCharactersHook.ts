import { useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { APP_PATH } from '../routes/constants';
import { useStateContext } from '../state';
import { ActionType } from '../state/actions';

export const useCharactersHook = () => {
    const { dispatch } = useStateContext();
    const navigate = useNavigate();

    const onFavoriteToggle = useCallback(
        (id: string) => {
            dispatch({
                type: ActionType.SET_FAV_CHARACTERS,
                payload: id,
            });
        },
        [dispatch]
    );

    const onDetailClick = useCallback(
        (id: string) => {
            navigate(`${APP_PATH.DETAILS}/${id}`);
        },
        [navigate]
    );

    return {
        onFavoriteToggle,
        onDetailClick,
    };
};
