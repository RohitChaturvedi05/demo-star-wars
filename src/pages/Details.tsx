import StarFilled from '@mui/icons-material/Star';
import StarBorderOutlined from '@mui/icons-material/StarBorderOutlined';
import {
    Box,
    Button,
    Card,
    CardContent,
    Container,
    Divider,
    Grid,
    Paper,
    Toolbar,
    Typography,
} from '@mui/material';
import { useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { Chip } from '../components/Chip';
import { ContentList } from '../components/ContentList';
import { EditableChip } from '../components/EditableChip';
import { Gender } from '../components/Gender';
import { Loader } from '../components/Loader';
import { useCharacterDetails } from '../hooks/useCharacterDetails';

const testid = 'character-details';
export const Details = () => {
    const { id } = useParams<{ id: string }>();

    const {
        message,
        character,
        planets,
        filmsNames,
        starShipsNames,
        isFavorite,
        onGenderChange,
        onAddToFavorites,
        getCharacterDetails,
    } = useCharacterDetails();

    useEffect(() => {
        if (id) getCharacterDetails(id);
        // need to initialize the state only once at the mount
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    return (
        <>
            <Toolbar />
            {!character && <Loader />}
            {character && (
                <Container maxWidth="lg" sx={{ py: 4 }} data-testid={testid}>
                    {message && <span>{message}</span>}
                    <Card elevation={3}>
                        <Box
                            sx={{
                                bgcolor: 'primary.main',
                                p: 3,
                                color: 'white',
                                display: 'flex',
                                justifyContent: 'space-between',
                                alignItems: 'center',
                            }}
                        >
                            <Typography
                                variant="h4"
                                component="h1"
                                data-testid={`${testid}$name`}
                            >
                                {character.name}
                            </Typography>
                            <Button
                                data-testid={`${testid}$fav-btn`}
                                variant="contained"
                                color={isFavorite ? 'secondary' : 'primary'}
                                onClick={() => onAddToFavorites(character.id)}
                                startIcon={
                                    isFavorite ? (
                                        <StarFilled />
                                    ) : (
                                        <StarBorderOutlined />
                                    )
                                }
                                sx={{
                                    bgcolor: 'white',
                                    color: 'primary.main',
                                    '&:hover': {
                                        bgcolor: 'grey.100',
                                    },
                                }}
                            >
                                Add/Remove Favorites
                            </Button>
                        </Box>
                        <CardContent sx={{ p: 4 }}>
                            <Grid container spacing={4}>
                                <Grid item xs={12} md={6}>
                                    <Typography
                                        variant="h6"
                                        gutterBottom
                                        color="primary"
                                    >
                                        Character Details
                                    </Typography>
                                    <Paper
                                        data-testid={`${testid}$details`}
                                        elevation={0}
                                        sx={{ p: 3, bgcolor: 'grey.50' }}
                                    >
                                        <Grid container spacing={2}>
                                            <Grid item xs={6}>
                                                <Chip
                                                    value={character.hair_color}
                                                    label="Hair Color"
                                                />
                                            </Grid>
                                            <Grid item xs={6}>
                                                <Chip
                                                    value={character.eye_color}
                                                    label="Eye Color"
                                                />
                                            </Grid>
                                            <Grid item xs={6}>
                                                <Gender
                                                    value={character.gender}
                                                    onGenderChange={(
                                                        value: Gender
                                                    ) =>
                                                        onGenderChange(
                                                            value,
                                                            character.id
                                                        )
                                                    }
                                                />
                                            </Grid>
                                            <Grid item xs={6}>
                                                <EditableChip
                                                    label="Height"
                                                    value={character.height}
                                                />
                                            </Grid>
                                            <Grid item xs={6}>
                                                <Chip
                                                    value={
                                                        planets[
                                                            character.homeworld
                                                        ]?.name ?? 'Unknown'
                                                    }
                                                    label="Home Planet"
                                                />
                                            </Grid>
                                        </Grid>
                                    </Paper>
                                </Grid>
                                <Grid item xs={12} md={6}>
                                    <Typography
                                        variant="h6"
                                        gutterBottom
                                        color="primary"
                                    >
                                        Related Content
                                    </Typography>
                                    <Paper
                                        elevation={0}
                                        sx={{
                                            p: 1,
                                            bgcolor: 'grey.50',
                                        }}
                                    >
                                        <ContentList
                                            data-testid={`${testid}$films`}
                                            label="Films"
                                            data={filmsNames}
                                        />
                                        <Divider sx={{ my: 2 }} />
                                        <ContentList
                                            data-testid={`${testid}$star-ships`}
                                            label="StarShips"
                                            data={starShipsNames}
                                        />
                                    </Paper>
                                </Grid>
                            </Grid>
                        </CardContent>
                    </Card>
                </Container>
            )}
        </>
    );
};

export default Details;
