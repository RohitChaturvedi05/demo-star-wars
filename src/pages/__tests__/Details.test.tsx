import { fireEvent, screen } from '@testing-library/react';
import {
    mockCharacterDetails,
    mockFilms,
    mockPlanets,
    mockStarship,
} from '../../__mocks__/charactersData';
import { renderWithState } from '../../__mocks__/renderWithState';
import { useCharacterDetails } from '../../hooks/useCharacterDetails';
import Details from '../Details';

// Acceptance Criteria - Character details view
// o display: name, hair colour, eye colour, gender, and home planet
// o list the films that the character has appeared in
// o list the starships that the character has piloted
// o provide the ability to add the character to the favourites list if they arenâ€™t already on it

jest.mock('react-router-dom', () => ({
    useNavigate: jest.fn(),
    useParams: jest.fn().mockReturnValue({ id: '1' }),
}));
jest.mock('../../hooks/useCharacterDetails');

const MockUseCharactersDetailsData = {
    character: mockCharacterDetails,
    films: [...mockFilms],
    starShips: mockStarship,
    planets: mockPlanets,
    filmsNames: [{ name: mockFilms[0].title }, { name: mockFilms[1].title }],
    starShipsNames: [
        { name: mockStarship[0].name },
        { name: mockStarship[1].name },
    ],
    isFavorite: false,
    onAddToFavorites: jest.fn(),
    getCharacterDetails: jest.fn(),
};

describe('pages/Details', () => {
    beforeEach(() => {
        (useCharacterDetails as jest.Mock).mockReturnValue(
            MockUseCharactersDetailsData
        );
    });

    it('should render without crashing', () => {
        const { container } = renderWithState(<Details />);
        expect(container).toMatchSnapshot();
    });

    it('should render character details with name, gender and home planet', () => {
        renderWithState(<Details />);

        expect(screen.getByText(mockCharacterDetails.name)).toBeInTheDocument();
        expect(
            screen.getByText(mockCharacterDetails.gender)
        ).toBeInTheDocument();
        expect(
            screen.getByText(mockCharacterDetails.eye_color)
        ).toBeInTheDocument();
        expect(
            screen.getByText(mockCharacterDetails.hair_color)
        ).toBeInTheDocument();
        expect(screen.getByText('Tatooine')).toBeInTheDocument();
    });
    it('should render list of films', () => {
        renderWithState(<Details />);
        expect(screen.getByText('Films')).toBeInTheDocument();
        MockUseCharactersDetailsData.filmsNames.forEach(({ name }) => {
            expect(screen.getByText(name)).toBeInTheDocument();
        });
    });
    it('should render list of starships', () => {
        renderWithState(<Details />);
        expect(screen.getByText('StarShips')).toBeInTheDocument();
        MockUseCharactersDetailsData.starShipsNames.forEach(({ name }) => {
            expect(screen.getByText(name)).toBeInTheDocument();
        });
    });

    it('should add/remove from favorites', () => {
        renderWithState(<Details />);
        const favBtn = screen.getByRole('button');
        expect(favBtn).toBeInTheDocument();
        fireEvent.click(favBtn);
        expect(
            MockUseCharactersDetailsData.onAddToFavorites
        ).toHaveBeenCalledWith(mockCharacterDetails.id);
    });
});
