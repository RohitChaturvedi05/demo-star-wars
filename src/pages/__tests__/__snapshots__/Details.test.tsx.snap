// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Details should render without crashing 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
  />
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-xsov46-MuiContainer-root"
    data-testid="character-details"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
      style="--Paper-shadow: 0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiBox-root css-1etdcjb"
      >
        <h1
          class="MuiTypography-root MuiTypography-h4 css-f84ov1-MuiTypography-root"
          data-testid="character-details$name"
        >
          Luke Skywalker
        </h1>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1wvlkvr-MuiButtonBase-root-MuiButton-root"
          data-testid="character-details$fav-btn"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1sh91j5-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="StarBorderOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m22 9.24-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28z"
              />
            </svg>
          </span>
          Add/Remove Favorites
        </button>
      </div>
      <div
        class="MuiCardContent-root css-21mhzp-MuiCardContent-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-4z7gnm-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1eoqm5i-MuiTypography-root"
            >
              Character Details
            </h6>
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-iaflhk-MuiPaper-root"
              data-testid="character-details$details"
              style="--Paper-shadow: none;"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-11nsb57-MuiGrid-root"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2 css-1x0mq5c-MuiTypography-root"
                  >
                    Hair Color
                  </h6>
                  <div
                    class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-16e8dsm-MuiChip-root"
                  >
                    <span
                      class="MuiChip-label MuiChip-labelMedium css-1dybbl5-MuiChip-label"
                    >
                      blond
                    </span>
                  </div>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2 css-1x0mq5c-MuiTypography-root"
                  >
                    Eye Color
                  </h6>
                  <div
                    class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-16e8dsm-MuiChip-root"
                  >
                    <span
                      class="MuiChip-label MuiChip-labelMedium css-1dybbl5-MuiChip-label"
                    >
                      blue
                    </span>
                  </div>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2 css-1x0mq5c-MuiTypography-root"
                  >
                    Gender
                  </h6>
                  <div
                    class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-16e8dsm-MuiChip-root"
                  >
                    <span
                      class="MuiChip-label MuiChip-labelMedium css-1dybbl5-MuiChip-label"
                    >
                      male
                    </span>
                  </div>
                  <button
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1w0xhdl-MuiButtonBase-root-MuiIconButton-root"
                    data-testid="btn-gender-edit"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
                      data-testid="EditIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                      />
                    </svg>
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2 css-1x0mq5c-MuiTypography-root"
                  >
                    Height
                  </h6>
                  <div
                    class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-16e8dsm-MuiChip-root"
                  >
                    <span
                      class="MuiChip-label MuiChip-labelMedium css-1dybbl5-MuiChip-label"
                    >
                      172
                    </span>
                  </div>
                  <button
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1w0xhdl-MuiButtonBase-root-MuiIconButton-root"
                    data-testid="btn-gender-edit"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
                      data-testid="EditIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                      />
                    </svg>
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2 css-1x0mq5c-MuiTypography-root"
                  >
                    Home Planet
                  </h6>
                  <div
                    class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-16e8dsm-MuiChip-root"
                  >
                    <span
                      class="MuiChip-label MuiChip-labelMedium css-1dybbl5-MuiChip-label"
                    >
                      Tatooine
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1eoqm5i-MuiTypography-root"
            >
              Related Content
            </h6>
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-s5vmot-MuiPaper-root"
              style="--Paper-shadow: none;"
            >
              <div
                data-testid="character-details$films"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
                >
                  Films
                </h6>
                <ul
                  class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
                >
                  <li
                    class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-cfd5wn-MuiListItem-root"
                  >
                    <div
                      class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
                    >
                      <span
                        class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                      >
                        A New Hope
                      </span>
                    </div>
                  </li>
                  <li
                    class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-cfd5wn-MuiListItem-root"
                  >
                    <div
                      class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
                    >
                      <span
                        class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                      >
                        The Empire Strikes Back
                      </span>
                    </div>
                  </li>
                </ul>
              </div>
              <hr
                class="MuiDivider-root MuiDivider-fullWidth css-zzw4lh-MuiDivider-root"
              />
              <div
                data-testid="character-details$star-ships"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
                >
                  StarShips
                </h6>
                <ul
                  class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
                >
                  <li
                    class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-cfd5wn-MuiListItem-root"
                  >
                    <div
                      class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
                    >
                      <span
                        class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                      >
                        Millennium Falcon
                      </span>
                    </div>
                  </li>
                  <li
                    class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-cfd5wn-MuiListItem-root"
                  >
                    <div
                      class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
                    >
                      <span
                        class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                      >
                        X-wing
                      </span>
                    </div>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
